polkadot/node/network/collator-protocol/src/validator_side_experimental/state.rs

```diff
@@ -14,19 +14,550 @@
+		&mut self,
+		sender: &mut Sender,
+		new_view: OurView,
+	) -> FatalResult<()> {
+		gum::trace!(
+			target: LOG_TARGET,
+			?new_view,
+			"Handling our view change",
+		);
+		let old_assignments = self.collation_manager.assignments();
+
+		self.collation_manager.update_view(sender, new_view).await?;
+
+		let new_assignments = self.collation_manager.assignments();
+		gum::trace!(
+			target: LOG_TARGET,
+			?old_assignments,
+			?new_assignments,
+			"Old assignments vs new assignments",
+		);
+
+		let maybe_disconnected_peers =
+			self.peer_manager.scheduled_paras_update(sender, new_assignments).await;
+
+		if !maybe_disconnected_peers.is_empty() {
+			gum::trace!(
+				target: LOG_TARGET,
+				?maybe_disconnected_peers,
+				"Disconnecting peers due to our view change",
+			);
+		}
+
+		self.collation_manager.remove_peers(maybe_disconnected_peers.iter());
+
+		Ok(())
+	}
+
+	/// Handle a finalized block notification.
+	pub async fn handle_finalized_block<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		hash: Hash,
+		number: BlockNumber,
+	) -> FatalResult<()> {
+		gum::trace!(
+			target: LOG_TARGET,
+			?hash,
+			number,
+			"Processing new block finality notification",
+		);
+
+		// Process reputation bumps
+		if let Err(err) = self
+			.peer_manager
+			.update_reputations_on_new_finalized_block(sender, (hash, number))
+			.await
+		{
+			err.split()?.log();
+		}
+
+		// Process potential changes in the registered paras set.
+		let session_index = match recv_runtime(request_session_index_for_child(hash, sender).await)
+			.await
+			.map_err(Error::Runtime)
+		{
+			Ok(session_index) => session_index,
+			Err(err) => {
+				err.split()?.log();
+				return Ok(())
+			},
+		};
+
+		self.peer_manager.prune_registered_paras(sender, session_index, hash).await;
+
+		Ok(())
+	}
+
+	/// Handle a new advertisement.
+	pub async fn handle_advertisement<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		peer_id: PeerId,
+		relay_parent: Hash,
+		maybe_prospective_candidate: Option<ProspectiveCandidate>,
+	) {
+		gum::debug!(
+			target: LOG_TARGET,
+			?relay_parent,
+			?maybe_prospective_candidate,
+			?peer_id,
+			"Received advertisement",
+		);
+
+		let Some(PeerInfo { state, .. }) = self.peer_manager.peer_info(&peer_id) else {
+			gum::warn!(
+				target: LOG_TARGET,
+				?relay_parent,
+				?peer_id,
+				?maybe_prospective_candidate,
+				"Received an advertisement from an unconnected peer"
+			);
+			return
+		};
+
+		// Advertised without being declared. Not a big waste of our time, so ignore it.
+		let PeerState::Collating(para_id) = state else {
+			gum::debug!(
+				target: LOG_TARGET,
+				?relay_parent,
+				?maybe_prospective_candidate,
+				?peer_id,
+				"Received advertisement for undeclared peer",
+			);
+			return
+		};
+
+		// We have a result here, but it's not worth affecting reputations because advertisements
+		// are cheap and quickly triaged.
+		match self
```

Comment:
This verification actually involves two other subsystems. So not super cheap, but should be cheap enough.